### PR TITLE
Add documentation for Træfik on docker swarm mode

### PR DESCRIPTION
**Pull request by vdemeester**
Still a little bit work in progress, I need to add output of commands 👼.

Also small update on the swarm one.

/cc @containous/traefik 

🐸

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

### CODE REVIEW

**Comment by bvis** [comment]
Hi, I've tested this docs with the 1.1.0-rc1 and it is not correct.

In you examples you are missing the needed labels to let traefik check the services. I've tried this instead with success:

```
docker network create traefik-net --driver overlay

docker service create \
  --name traefik \
  --constraint=node.role==manager \
  --publish 80:80 --publish 8080:8080 \
  --mount type=bind,source=/var/run/docker.sock,target=/var/run/docker.sock \
  --network traefik-net \
  traefik:camembert \
  --docker \
  --docker.swarmmode \
  --docker.domain=docker.localhost \
  --docker.watch \
  --logLevel=DEBUG \
  --web

docker service create \
    --name whoami0 \
    --network traefik-net \
    --label "traefik.backend=whoami" \
    --label "traefik.frontend.rule=Host:whoami.docker.localhost" \
    --label "traefik.port=80" \
    emilevauge/whoami

docker service create \
    --name whoami1 \
    --network traefik-net \
    --label "traefik.backend=whoami" \
    --label "traefik.frontend.rule=Host:whoami.docker.localhost" \
    --label "traefik.port=80" \
    emilevauge/whoami

$ curl -H Host:whoami.docker.localhost http://localhost
Hostname: 640be50494d2
IP: 127.0.0.1
IP: ::1
IP: 10.0.6.7
IP: 10.0.6.6
IP: fe80::42:aff:fe00:607
IP: 172.21.0.4
IP: fe80::42:acff:fe15:4
GET / HTTP/1.1
Host: whoami.docker.localhost
User-Agent: curl/7.43.0
Accept: */*
Accept-Encoding: gzip
X-Forwarded-For: 10.255.0.4
X-Forwarded-Host: whoami.docker.localhost
X-Forwarded-Proto: http
X-Forwarded-Server: 57dd126575c5

 $ curl -H Host:whoami.docker.localhost http://localhost
Hostname: c5c321cac154
IP: 127.0.0.1
IP: ::1
IP: 10.0.6.5
IP: 10.0.6.4
IP: fe80::42:aff:fe00:605
IP: 172.21.0.3
IP: fe80::42:acff:fe15:3
GET / HTTP/1.1
Host: whoami.docker.localhost
User-Agent: curl/7.43.0
Accept: */*
Accept-Encoding: gzip
X-Forwarded-For: 10.255.0.4
X-Forwarded-Host: whoami.docker.localhost
X-Forwarded-Proto: http
X-Forwarded-Server: 57dd126575c5
```

File: docs/user-guide/swarm-mode.md [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+User-Agent: curl/7.35.0
+Accept: */*
+Accept-Encoding: gzip
+X-Forwarded-For: 192.168.99.1
+X-Forwarded-Host: 10.0.9.4:80
+X-Forwarded-Proto: http
+X-Forwarded-Server: 8fbc39271b4c
+```
+
+Note that as Træfik is publish, you can access it from any machine
```
Comment (ralphtheninja) [inline docs/user-guide/swarm-mode.md]: is published

File: docs/user-guide/swarm-mode.md [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+(and not only the manager).
```
Comment (ralphtheninja) [inline docs/user-guide/swarm-mode.md]: remove parenthesis

File: docs/user-guide/swarm.md [hunk 3]
```diff
@@ -1,7 +1,7 @@
 # Swarm cluster
 
 This section explains how to create a multi-host [swarm](https://docs.docker.com/swarm) cluster using [docker-machine](https://docs.docker.com/machine/) and how to deploy Træfɪk on it.
-The cluster will be made of:
+The cluster is be made of:
```
Comment (ralphtheninja) [inline docs/user-guide/swarm.md]: The cluster consists of

File: docs/user-guide/swarm-mode.md [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+
+Let's explain this command:
+
+- `--publish 80:80 --publish 8080:8080`: we publish port `80` and
+  `8080` on the cluster.
+- `--constraint=node.role==manager`: we ask docker to schedule Træfik
+  on a manager node.
+- `--mount type=bind,source=/var/run/docker.sock,target=/var/run/docker.sock`:
+  we bind mount the docker socket where Træfik is scheduled to be able
+  to speak to him.
```
Comment (ralphtheninja) [inline docs/user-guide/swarm-mode.md]: to speak to the docker daemon

**Comment by weitzj** [comment]
It would be nice to have 2 separate networks for the `whoami` backends. 
I wilk probably add an example with traefik, docker-compose bundle, docker service deploy where separate networks are even more important

**Comment by vdemeester** [comment]
@ralphtheninja thanks :angel: 

@bvis Yes, I forgot `--label "traefik.port=80"` in my example, thanks :angel:. I can make an example of what you describe but that's not what I was going for here. I'm using the same image twice because I'm lazy :joy: but I really want 2 different service served by Træfik as they were different app.

@weitzj I definitely went for the simplest example ever here :sweat_smile:. We could enhance the documentation with multiple network (one for each app served) and a way to update the `traefik` service to connect to those as needed (but doesn't look like there is a `docker service update --add-network` yet though :sweat:).

**Comment by emilevauge** [comment]
Needs a rebase :)

**Comment by tlvenn** [comment]
Unless I am mistaken there is a slight error with how the placement contraint is declared, the official syntax is `--constraint string` not `--constraint=string`
> https://docs.docker.com/engine/reference/commandline/service_create/#/specify-service-constraints---constraint

**Comment by vdemeester** [comment]
@tlvenn both should work the same 👼

**Comment by tlvenn** [comment]
Ya I didnt check but figured it would however I believe it would be a little bit better to stick with the official syntax, especially given that's the one you are already using for the other options ;)

Anyway, I am nitpicking, thanks in any case, for putting the doc together.

**Comment by vdemeester** [comment]
:stuck_out_tongue_closed_eyes:

**Comment by tlvenn** [comment]
One new question, did you investigate the `--endpoint-mode` option ? By default docker will use and create a virtual ip to load balance itself the service. One can use `--endpoint-mode dnsrr` to disable VIP and rely on simple dns round robin. Their documentation seems to suggest to use this when using your own load balancer, which is what we are doing.
> You can configure the service to use DNS round-robin directly without using a VIP, by setting the --endpoint-mode dnsrr when you create the service. DNS round-robin is useful in cases where you want to use your own load balancer.

Should we update the doc to reflect that ?

**Comment by ralphtheninja** [comment]
> Should we update the doc to reflect that ?

Make a PR and lets discuss _that_ instead ;)

**Comment by tlvenn** [comment]
Well there is no point to create a PR if we dont even agree it should be amended...
The doc works perfectly fine as it is, it just create an un necessary IP for the service being load balanced by Traefik.

In the end, there are 3 choices:

1. Keep it simple and dont mention `--endpoint-mode dnsrr`
2. Amend the doc to use `--endpoint-mode dnsrr` and explain it.
3. Amend the doc to mention `--endpoint-mode dnsrr` as an option to avoid creating the VIP.

In case we choose 2 or 3, I am happy to create the PR.

**Comment by tlvenn** [comment]
One last comment, for security purpose, the docker socket should really be mounted in read only mode so that in the unlikely event the container is compromised, it does not mean the whole cluster has been taken over...

**Pull request merged**